byterun/major_gc.c

```diff
@@ -169,11 +156,8 @@ static void mark_slice (intnat work)
           child = Field (v, i);
 #ifdef NATIVE_CODE_AND_NO_NAKED_POINTERS
           if (Is_block (child)
-                && Wosize_val (child) > 0  /* Atoms never need to be marked. */
-                /* Closure blocks contain code pointers at offsets that cannot
-                   be reliably determined, so we always use the page table when
-                   marking such values. */
-                && (!marking_closure || Is_in_heap (child))) {
+                /* Atoms never need to be marked. */
+                && Wosize_val (child) > 0) {
```

Comment:
Why not make atom headers black and get rid of this test? Then you can use `Nnp_or_is_in_heap` and remove the `#ifdef` entirely.